File: ffi/snapshotclient.go
package ffi

import (
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// SnapshotErrorKind enumerates the snapshot RPC error codes.
type SnapshotErrorKind string

const (
	SnapshotPlanInvalid            SnapshotErrorKind = "state_sync_plan_invalid"
	SnapshotMetadataMismatch       SnapshotErrorKind = "state_sync_metadata_mismatch"
	SnapshotProofEncodingInvalid   SnapshotErrorKind = "state_sync_proof_encoding_invalid"
	SnapshotVerificationIncomplete SnapshotErrorKind = "state_sync_verification_incomplete"
	SnapshotVerifierIO             SnapshotErrorKind = "state_sync_verifier_io"
	SnapshotPipelineError          SnapshotErrorKind = "state_sync_pipeline_error"
	SnapshotPrunerStateError       SnapshotErrorKind = "state_sync_pruner_state_error"
	SnapshotTransport              SnapshotErrorKind = "transport"
	SnapshotUnknown                SnapshotErrorKind = "unknown"
)

// SnapshotError wraps RPC failures with typed context.
type SnapshotError struct {
	Status     int
	Message    string
	Kind       SnapshotErrorKind
	RetryAfter time.Duration
}

func (e *SnapshotError) Error() string {
	return strings.TrimSpace(e.Message)
}

func parseRetryDelay(resp *http.Response, minBackoff time.Duration) time.Duration {
	if reset := resp.Header.Get("X-RateLimit-Reset"); reset != "" {
		if secs, err := strconv.Atoi(reset); err == nil && secs >= 0 {
			if d := time.Duration(secs) * time.Second; d > 0 {
				return d
			}
		}
	}
	if after := resp.Header.Get("Retry-After"); after != "" {
		if secs, err := strconv.Atoi(after); err == nil && secs >= 0 {
			d := time.Duration(secs) * time.Second
			if d > 0 {
				return d
			}
		}
	}
	return minBackoff
}

// ClassifySnapshotResponse turns an HTTP response into a typed error when the
// server reports a snapshot failure. Successful responses pass through.
func ClassifySnapshotResponse(resp *http.Response, minBackoff time.Duration) error {
	if resp.StatusCode < 400 {
		return nil
	}

	body, _ := io.ReadAll(resp.Body)
	_ = resp.Body.Close()

	var envelope struct {
		Error string  `json:"error"`
		Code  *string `json:"code"`
	}
	_ = json.Unmarshal(body, &envelope)

	kind := SnapshotUnknown
	if envelope.Code != nil {
		switch *envelope.Code {
		case string(SnapshotPlanInvalid):
			kind = SnapshotPlanInvalid
		case string(SnapshotMetadataMismatch):
			kind = SnapshotMetadataMismatch
		case string(SnapshotProofEncodingInvalid):
			kind = SnapshotProofEncodingInvalid
		case string(SnapshotVerificationIncomplete):
			kind = SnapshotVerificationIncomplete
		case string(SnapshotVerifierIO):
			kind = SnapshotVerifierIO
		case string(SnapshotPipelineError):
			kind = SnapshotPipelineError
		case string(SnapshotPrunerStateError):
			kind = SnapshotPrunerStateError
		default:
			kind = SnapshotUnknown
		}
	}

	return &SnapshotError{
		Status:     resp.StatusCode,
		Message:    envelope.Error,
		Kind:       kind,
		RetryAfter: parseRetryDelay(resp, minBackoff),
	}
}

// SnapshotGET executes an HTTP GET and maps snapshot RPC errors to typed failures.
func SnapshotGET(client *http.Client, url string, minBackoff time.Duration) error {
	resp, err := client.Get(url)
	if err != nil {
		return &SnapshotError{
			Status:     http.StatusServiceUnavailable,
			Message:    err.Error(),
			Kind:       SnapshotTransport,
			RetryAfter: minBackoff,
		}
	}
	defer resp.Body.Close()

	if err := ClassifySnapshotResponse(resp, minBackoff); err != nil {
		return err
	}
	return nil
}

// IsRetryable returns whether the error represents a transient snapshot condition.
func (e *SnapshotError) IsRetryable() bool {
	switch e.Kind {
	case SnapshotProofEncodingInvalid, SnapshotVerificationIncomplete, SnapshotVerifierIO, SnapshotPipelineError:
		return true
	default:
		return false
	}
}

// Ensure SnapshotError matches the standard error interface.
var _ error = (*SnapshotError)(nil)

var (
	ErrMissingClient = errors.New("http client is nil")
)

// SnapshotRequest validates the client and performs a GET.
func SnapshotRequest(client *http.Client, url string, minBackoff time.Duration) error {
	if client == nil {
		return ErrMissingClient
	}
	return SnapshotGET(client, url, minBackoff)
}

File: validator-ui/src/lib/snapshotClient.ts
export type SnapshotErrorCode =
  | "state_sync_plan_invalid"
  | "state_sync_metadata_mismatch"
  | "state_sync_proof_encoding_invalid"
  | "state_sync_verification_incomplete"
  | "state_sync_verifier_io"
  | "state_sync_pipeline_error"
  | "state_sync_pruner_state_error"
  | "transport"
  | "unknown";

export class SnapshotError extends Error {
  status: number;
  code: SnapshotErrorCode;
  retryAfter: number;

  constructor(message: string, status: number, code: SnapshotErrorCode, retryAfter: number) {
    super(message);
    this.name = "SnapshotError";
    this.status = status;
    this.code = code;
    this.retryAfter = retryAfter;
  }

  get retryable(): boolean {
    return (
      this.code === "state_sync_proof_encoding_invalid" ||
      this.code === "state_sync_verification_incomplete" ||
      this.code === "state_sync_verifier_io" ||
      this.code === "state_sync_pipeline_error" ||
      this.code === "transport"
    );
  }
}

function parseRetryAfter(headers: Headers, minDelayMs: number): number {
  const resetHeader = headers.get("X-RateLimit-Reset");
  if (resetHeader) {
    const asNumber = Number.parseInt(resetHeader, 10);
    if (!Number.isNaN(asNumber) && asNumber > 0) {
      return asNumber * 1000;
    }
  }

  const retryAfter = headers.get("Retry-After");
  if (retryAfter) {
    const asNumber = Number.parseInt(retryAfter, 10);
    if (!Number.isNaN(asNumber) && asNumber > 0) {
      return asNumber * 1000;
    }
  }

  return minDelayMs;
}

function mapCode(code?: string | null): SnapshotErrorCode {
  switch (code) {
    case "state_sync_plan_invalid":
    case "state_sync_metadata_mismatch":
    case "state_sync_proof_encoding_invalid":
    case "state_sync_verification_incomplete":
    case "state_sync_verifier_io":
    case "state_sync_pipeline_error":
    case "state_sync_pruner_state_error":
      return code;
    default:
      return code ? "unknown" : "unknown";
  }
}

export async function snapshotRequest(
  input: RequestInfo | URL,
  init: RequestInit = {},
  minDelayMs = 50,
): Promise<Response> {
  try {
    const response = await fetch(input, init);
    if (response.ok) {
      return response;
    }

    const retryAfter = parseRetryAfter(response.headers, minDelayMs);
    try {
      const payload = (await response.json()) as { error?: string; code?: string };
      throw new SnapshotError(payload.error ?? "snapshot RPC failed", response.status, mapCode(payload.code), retryAfter);
    } catch (err) {
      if (err instanceof SnapshotError) {
        throw err;
      }
      const text = await response.text();
      throw new SnapshotError(text || "snapshot RPC failed", response.status, "unknown", retryAfter);
    }
  } catch (err) {
    if (err instanceof SnapshotError) {
      throw err;
    }
    throw new SnapshotError((err as Error).message, 503, "transport", minDelayMs);
  }
}
